packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -1038,109 +971,67 @@ contract AssetManager is ISemver, IERC721Receiver, IAssetManager {
 
         unchecked {
+            // asset
             vault.asset.totalKroShares -= kroShares;
-            vault.asset.totalKghShares -= kghShares;
-            delete vault.kghDelegators[owner].shares[tokenId];
-
-            uint128 kroInKgh = KGH_MANAGER.totalKroInKgh(tokenId);
-            uint128 baseRewardsToReceive = kroAssetsToWithdraw - kroInKgh;
-            boostedRewardsToReceive = kghAssetsToWithdraw - VKRO_PER_KGH;
-
-            vault.asset.totalKro -= kroAssetsToWithdraw;
-            vault.asset.totalKroInKgh -= kroInKgh;
             vault.asset.totalKgh -= 1;
-            vault.asset.boostedReward -= boostedRewardsToReceive;
-
-            KghDelegatorShares memory pendingShares = KghDelegatorShares({
-                kro: kroShares.mulDiv(baseRewardsToReceive, kroAssetsToWithdraw),
-                kgh: kghShares.mulDiv(boostedRewardsToReceive, kghAssetsToWithdraw)
-            });
-
-            _addPendingKghShares(
-                vault,
-                tokenIds,
-                baseRewardsToReceive,
-                boostedRewardsToReceive,
-                pendingShares
-            );
-        }
+            vault.asset.totalKro -= kroAssets;
+            vault.asset.totalKroInKgh -= kroInKgh;
 
-        if (kroAssetsToWithdraw + boostedRewardsToReceive > 0) {
-            VALIDATOR_MANAGER.updateValidatorTree(validator, true);
+            // delegator
+            kghDelegator.kghNum -= 1;
+            delete kghDelegator.delegatedAt[tokenId];
+            delete kghDelegator.kroShares[tokenId];
         }
     }
 
     /**
      * @notice Internal function to undelegate KGHs from the validator.
      *
      * @param validator           Address of the validator.
-     * @param tokenIds            Array of token ids of KGHs to undelegate.
+     * @param delegator           Address of the delegator.
+     * @param kghCount            The number of KGH token to undelegate.
      * @param kroInKghs           The amount of KRO in the KGHs.
+     * @param kroAssets           The amount of KRO Assets to undelegate.
      * @param kroShares           The amount of KRO shares to undelegate.
-     * @param kghShares           The amount of KGH shares to undelegate.
-     * @param kghAssetsToWithdraw The amount of KGH assets to withdraw.
      */
-    function _initUndelegateKghBatch(
+    function _undelegateKghBatch(
         address validator,
-        uint256[] calldata tokenIds,
+        address delegator,
+        uint128 kghCount,
         uint128 kroInKghs,
-        uint128 kroShares,
-        uint128 kghShares,
-        uint128 kghAssetsToWithdraw
+        uint128 kroAssets,
+        uint128 kroShares
     ) internal {
         Vault storage vault = _vaults[validator];
-        uint128 kroAssetsToWithdraw = _convertToKroAssets(validator, kroShares);
-        uint128 boostedRewardsToReceive;
+        KghDelegator storage kghDelegator = vault.kghDelegators[delegator];
```

Comment:
How about just declaring `Asset storage asset`? That will be better for gas usage I think, combined with the comment at L#1034.